src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,905 @@
+        System.arraycopy(arr2, 0, temp, arr1.length, arr2.length);
+        return temp;
+    }
+
+    /**
+     * Generates a key from password to get NTLMv2 hash Section 3.3.2 NTLM v2 Authentication
+     *
+     * @return NT response key
+     * @throws InvalidKeyException
+     */
+    private byte[] ntowfv2() throws InvalidKeyException {
+        return hmacMD5(context.passwordHash,
+                (null != context.upperUserName) ? unicode(context.upperUserName + context.domainName)
+                                                : unicode(context.domainName));
+    }
+
+    /**
+     * Computes the NT response and keys from the Challenge message
+     *
+     * Section 3.3.2 NTLM v2 Authentication
+     *
+     * <pre>
+     * Set temp to ConcatenationOf(Responserversion, HiResponserversion, 
+     * Z(6), Time, ClientChallenge, Z(4), ServerName, Z(4)) 
+     * Set NTProofStr to HMAC_MD5(ResponseKeyNT,  
+     * ConcatenationOf(CHALLENGE_MESSAGE.ServerChallenge,temp)) 
+     * Set NtChallengeResponse to ConcatenationOf(NTProofStr, temp) 
+     * Set LmChallengeResponse to ConcatenationOf(HMAC_MD5(ResponseKeyLM,  
+     * ConcatenationOf(CHALLENGE_MESSAGE.ServerChallenge, ClientChallenge)), 
+     * ClientChallenge )
+     * </pre>
+     *
+     * @param responseKeyNT
+     *        NT response hash key
+     * @return computed response
+     * @throws InvalidKeyException
+     * @throws NoSuchAlgorithmException
+     */
+    private byte[] computeResponse(final byte[] responseKeyNT) throws InvalidKeyException, NoSuchAlgorithmException {
+        // get random client challenge nonce
+        byte[] clientNonce = new byte[NTLM_CLIENT_NONCE_LENGTH];
+        (new Random()).nextBytes(clientNonce);
+
+        // get client challenge blob
+        byte[] temp = generateClientChallengeBlob(clientNonce);
+
+        byte[] ntProofStr = hmacMD5(responseKeyNT, concat(context.serverChallenge, temp));
+
+        context.sessionBaseKey = hmacMD5(responseKeyNT, ntProofStr);
+
+        return concat(ntProofStr, temp);
+    }
+
+    /**
+     * Generates the NT Challenge response to server Challenge
+     *
+     * Section 3.3.2 NTLM v2 Authentication
+     *
+     * @param clientNonce
+     *        client challenge nonce
+     * @return NT challenge response
+     * @throws InvalidKeyException
+     * @throws NoSuchAlgorithmException
+     */
+    private byte[] getNtChallengeResp() throws InvalidKeyException, NoSuchAlgorithmException {
+        byte[] responseKeyNT = ntowfv2();
+        return computeResponse(responseKeyNT);
+    }
+
+    /**
+     * Generates the Type 3 NTLM Authentication message
+     *
+     * Section 2.2.1.3 AUTHENTICATE_MESSAGE
+     *
+     * @return NTLM Authenticate message
+     * @throws SQLServerException
+     */
+    private byte[] generateNtlmAuthenticate() throws SQLServerException {
+        int domainNameLen = null != context.domainBytes ? context.domainBytes.length : 0;

```

Comment:
Might be worth making a helper function instead? Seems like this comparison logic is used a lot for these byte buffers.

```java
private int getByteArrayLength(byte[] b) {
    return null == b ? 0 : b.length();
}

int domainNameLen = getByteArrayLength(context.DomainBytes);
```